debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,452 @@
+    private final DataChangeEventListener dataListener;
+    private long totalRowsScanned = 0;
+    private final ReplicaSets replicaSets;
+    private final ConnectionContext connectionContext;
+    private final MongoDbTaskContext taskContext;
+
+    private MongoDbCollectionSchema currentCollection;
+
+    protected EventDispatcher<CollectionId> dispatcher;
+    protected IncrementalSnapshotContext<T> context = null;
+    protected final Map<Struct, Object[]> window = new LinkedHashMap<>();
+    private MongoPrimary primary;
+    private CollectionId signallingCollectionId;
+
+    public MongoDbIncrementalSnapshotChangeEventSource(MongoDbConnectorConfig config,
+                                                       MongoDbTaskContext taskContext,
+                                                       ReplicaSets replicaSets,
+                                                       EventDispatcher<CollectionId> dispatcher,
+                                                       MongoDbSchema collectionSchema,
+                                                       Clock clock,
+                                                       SnapshotProgressListener progressListener,
+                                                       DataChangeEventListener dataChangeEventListener) {
+        this.connectorConfig = config;
+        this.replicaSets = replicaSets;
+        this.taskContext = taskContext;
+        this.connectionContext = taskContext.getConnectionContext();
+        this.dispatcher = dispatcher;
+        this.collectionSchema = collectionSchema;
+        this.clock = clock;
+        this.progressListener = progressListener;
+        this.dataListener = dataChangeEventListener;
+        this.signallingCollectionId = connectorConfig.getSignalingDataCollectionId() == null ? null
+                : CollectionId.parse("UNUSED", connectorConfig.getSignalingDataCollectionId());
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void closeWindow(Partition partition, String id, OffsetContext offsetContext) throws InterruptedException {
+        context = (IncrementalSnapshotContext<T>) offsetContext.getIncrementalSnapshotContext();
+        if (!context.closeWindow(id)) {
+            return;
+        }
+        sendWindowEvents(partition, offsetContext);
+        readChunk();
+    }
+
+    protected String getSignalCollectionName(String dataCollectionId) {
+        return dataCollectionId;
+    }
+
+    protected void sendWindowEvents(Partition partition, OffsetContext offsetContext) throws InterruptedException {
+        LOGGER.debug("Sending {} events from window buffer", window.size());
+        offsetContext.incrementalSnapshotEvents();
+        for (Object[] row : window.values()) {
+            sendEvent(partition, dispatcher, offsetContext, row);
+        }
+        offsetContext.postSnapshotCompletion();
+        window.clear();
+    }
+
+    // TODO Used typed dispatcher and offset context
```

Comment:
What about this TODO? Can you log a follow-up issue if you don't plan to address it as part of this PR?

Reply:
Handled in follo-up Jira